merkledag/merkledag.go

```diff
@@ -255,14 +263,39 @@ func GetLinksWithDAG(ng ipld.NodeGetter) GetLinks {
 // unseen children to the passed in set.
 // TODO: parallelize to avoid disk latency perf hits?
 func EnumerateChildren(ctx context.Context, getLinks GetLinks, root *cid.Cid, visit func(*cid.Cid) bool) error {
+	visitDepth := func(c *cid.Cid, maxDepth int) bool {
+		return visit(c)
+	}
+	return EnumerateChildrenMaxDepth(ctx, getLinks, root, -1, visitDepth)
+}
+
+// EnumerateChildrenMaxDepth walks the dag below the given root to the given
+// depth. The root is at level 0, the children are at level 1 and so on.
+// Thus, setting depth to two, will walk the root, the children, and the
+// children of the children.
+// Setting depth to a negative number will walk the full tree.
+func EnumerateChildrenMaxDepth(ctx context.Context, getLinks GetLinks, root *cid.Cid, maxDepth int, visit func(*cid.Cid, int) bool) error {
```

Comment:
A more general way to do this (and the async version) would be to:

1. Make the visit function take a *current* depth. I'd *like* to pass a full path (as a `[]string`) but that might get expensive.
2. Have the visit function determine if we should go deeper.

Reply:
> Have the visit function determine if we should go deeper.

But this is what is does. You can either remember how far the item is to the limit (maxDepth, as it does now), or you can remeber how deep the item is and what the absolute depth limit is. The way it's done now requires only remembering one thing.